Hallucinations Overhaul

Doxxmedearly opened the pull request:
A long-overdue rehaul of "help monkey punch me."

Ports most of hallucination framework from bay and expands it tremendously. It's not a total port but it's way cleaner than the ages-old hallucination code, and so easier to expand on. Removed lots of the old effects and expanded on the ones I kept. 

Huge lists got shoved into text files: hallucinated_phrases (Things you imagine others say), hallucinated_thoughts (Thought prompts that hit you), hallucinated_actions (things you imagine other people doing).

I have two main goals:
1) Provide a huge variety of things that can happen to a person, to mitigate it getting predictable/stale as much as I can.
2) Use the effects to facilitate RP. 

Pun-pun punching you, 'fake death', and weird HUD updates were more like LRP interruptions to a person's RP. Now there's a variety of things for people to work with. Variety goes a long way beyond what I've written, too; even getting the same prompt can create different RP depending on when you get it, as what character, who's around them, etc. Many of the new effects involve the people around you.

Tidying up how chems like mental medications affect hallucinations will come in a followup PR, but anti-hallucination meds in their current state will still help. Replacing fake pun-pun is also still in the works, but a larger undertaking and would significantly increase the size of this PR.

TODO:
- [x]  Get passive and colorblind hallucinations to work
- [x]  Get fake mind read verb to work
- [x]  Get "talking" to display names properly and loop properly
- [x]  Sound and Mirage variety
- [x]  Finalize txt lists by editing, adding, removing some options.
- [x]  Finalize txt files' permanent home in the code, if it needs to be moved. I don't know if adding a folder to config is a sin help.
- [x]  Get hear_say, hear_radio, and mistyping messages in PDAs while hallucinating to work. Get the latter to log to admins when message changed.
- [x]  Finalize delays, durations, and power threshholds. Ensure a variety of effects are available.
- [x]  Test to hell and back
- [x]  Self reminder to remove debug messages (can recover:)
- [x] Verb issues

Doxxmedearly commented:
!wip

Doxxmedearly commented:
!review

Doxxmedearly commented:
Ready for review for now, though I'm still doing plenty of testing and finalization of the other checkmarks. If someone could hit me with a DNM that would be great. I'm at the point where I'll need more experienced input; the effects are working currently.

bls be gentle, I know there's lots of spaghet to clean up. If you ask "why don't you use x instead" the answer is I probably didn't know about it, pls grant me your knowledge. except span. I'll merge master and get to using that new macro.

in particular i don't know if the message things on hear_say, hear_radio, and fake PDA message need to be logged to admins? I feel like it might be prudent to have as evidence for any complaints that may arise. Let me know if I did the logging correctly, or if anything else should be logged.

also unsure if the txt lists can be placed where I placed them in config, or where they should go. Advice appreciated.

Doxxmedearly commented:
!wip

Doxxmedearly commented:
!review

NonQueueingMatt commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -1116,6 +1116,12 @@ var/global/list/obj/item/device/pda/PDAs = list()
+		if(C.hallucination >= 50 && prob(C.hallucination / 10))			//If you're really hallucinating, you might not be typing what you think you are
+			var/t_orig = t
+			t = pick(hallucinated_phrases)								//see carbon/hallucination.dm
+			log_pda("[usr] (PDA: [src.name]) typed \"[t_orig]\" then hallucination changed it to \"[t]\". Recipient was [P.owner]",ckey=key_name(usr),ckey_target=key_name(P.owner))
```
```suggestion
			log_pda("[usr] (PDA: [C.name]) typed \"[t_orig]\" then hallucination changed it to \"[t]\". Recipient was [P.owner]",ckey=key_name(C),ckey_target=key_name(P.owner))
```

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+var/list/hallucinated_thoughts = file2list("config/hallucination/hallucinated_thoughts.txt")
+
+
+/mob/living/carbon/var/next_hallucination = 0			//Hallucination spam limit var
```
I don't quite like the declaration here. Should be like this:
````
/mob/living/carbon
    var/next_hallucination = 0 //yada
    var/list/hallucinations = list() //yada2
```
If you do this, people can find the variables by searching `carbon$` on VSC with regex.

NonQueueingMatt replied:
Also, just a thought, maybe this stuff should be human only and located in the brain, but I don't mind anyway.

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+		log_say("Hallucination level changed [orig_message] by [speaker] to [message] for [key_name(src)].", ckey=key_name(src))
+	..()
+/mob/living/carbon/hear_radio(var/message, var/verb="says", var/datum/language/language=null, var/part_a, var/part_b, var/mob/speaker = null, var/hard_to_hear = 0, var/vname ="")
+	if(hallucination >= 60 && prob(1))
```
```suggestion
	..()

/mob/living/carbon/hear_radio(var/message, var/verb="says", var/datum/language/language=null, var/part_a, var/part_b, var/mob/speaker = null, var/hard_to_hear = 0, var/vname ="")
	if(hallucination >= 60 && prob(1))
```

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+		message = LAZYPICK(hallucinated_phrases, message)
+		log_say("Hallucination level changed [orig_message] by [speaker] to [message] for [key_name(src)].", ckey=key_name(src))
+	..()
+/mob/living/carbon/hear_radio(var/message, var/verb="says", var/datum/language/language=null, var/part_a, var/part_b, var/mob/speaker = null, var/hard_to_hear = 0, var/vname ="")
```
```suggestion
/mob/living/carbon/hear_radio(var/message, var/verb="says", var/datum/language/language, var/part_a, var/part_b, var/mob/speaker, var/hard_to_hear = 0, var/vname ="")
```

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+	if(!hallucination)  //We're done
+		return
+
+	if(!client || stat || world.time < next_hallucination)
```
If they're unconscious their hallucinations don't tick?

Doxxmedearly replied:
Hallucination ticks down before the stat check, on line 25. It ticks down but they do not experience hallucination effects

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+	if(candidates.len)
+		var/datum/hallucination/H = pick(candidates)
+		H.holder = src
+		H.activate()
```
Yeah, no, don't do this every life tick. Make a new subsystem or use the brain trauma one to initialize a list of hallucinations at roundstart and then pick one from the list.

Doxxmedearly replied:
It won't do this candidate check on every life tick. It will only do it if the `if(!client || stat || world.time < next_hallucination)` check passes.

Doxxmedearly replied:
Additionally, it's impossible to just initialize a list at roundstart if we want these to be varied, since the can_affect() check is what checks for if there are mobs around, how badly they're hallucinating, if they're deaf, etc. which will give us candidates that can currently affect the person. However I will shove this checking and all of the lists into a subsystem since that's a way better idea.

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+
+/datum/hallucination/Destroy()
+	. = ..()
+	holder = null
```
Destroy() needs to do this:

```
/datum/hallucination/Destroy()
	holder = null
	return ..()
```

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+		hallucination_emote(holder)		//Always a chance to involuntarily emote to others as if on drugs
+		holder.hallucinations -= src
+	qdel(src)
+	holder = null
```
already handled in destroy

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+	var/hearing_dependent = FALSE	//Deaf people won't get these if true
+
+///////////Things you involuntarily emote while hallucinating //////////////
+	var/list/hal_emote = list("mutters quietly.", "stares.", "grunts.", "looks around.", "twitches.", "shivers.", "swats at the air.", "wobbles.", "gasps!", "blinks rapidly.", "murmurs.",
```
```suggestion
	var/static/list/hal_emote = list("mutters quietly.", "stares.", "grunts.", "looks around.", "twitches.", "shivers.", "swats at the air.", "wobbles.", "gasps!", "blinks rapidly.", "murmurs.",
```

NonQueueingMatt replied:
Also why is this stuff on the datum? This looks like a global list kinda thing. Maybe this should be in the same subsystem as the hallucination datum list.

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+				"dry heaves!", "twitches violently.", "giggles.", "drools.", "scratches all over.", "grinds their teeth.", "whispers something quietly.")
+
+/////////////// Sender list of names for fake messages/announcements/other uses. Some copied over from spam messages//////
+	var/list/message_sender = list("Mom", "Dad", "Captain", "Captain(as Captain)", "help", "Home", "MaxBet Online Casino", "IDrist Corp", "Dr. Maxman",
```
```suggestion
	var/static/list/message_sender = list("Mom", "Dad", "Captain", "Captain(as Captain)", "help", "Home", "MaxBet Online Casino", "IDrist Corp", "Dr. Maxman",
```

NonQueueingMatt replied:
Ditto above.

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+	var/duration = 10				//how long before we call end()
+	var/min_power = 0 				//mobs only get this hallucination at this threshold
+	var/max_power = INFINITY		//mobs don't get this hallucination if it's above this threshold. Used to weed out more common ones if you're super fucked up
+	var/hearing_dependent = FALSE	//Deaf people won't get these if true
```
Maybe this should be a bitflag and then you can have things like this:

`var/special_flags = HEARING_DEPENDANT | SOMETHING_ELSE_YOU_COME_UP_WITH_IN_THE_FUTURE`

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+
+
+/datum/hallucination/paranoia
+
```
No special settings here?

Doxxmedearly replied:
Nope. No duration since it's a one-shot thing and I want it to happen at all levels of hallucination.

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+	for(var/mob/living/carbon/human/H in living_mob_list)
+		if(H.client && !player_is_antag(H, only_offstation_roles = TRUE))	//adds current players to default list to provide variety. leaves out offstation antags.
+			sender += H
+	holder.show_message("<b>Message from [pick(sender)] to [holder.name] ([M.job]),</b> \"[pick(hallucinated_phrases)]\" (<FONT color = blue><u>reply</u></FONT>)")
```
Show_message? I'm sure this should be to_chat.

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+
+/datum/hallucination/pda/start()
+	var/list/sender = message_sender
+	var/mob/living/carbon/human/M = holder
```
This sounds unsafe. Typecheck it.

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+	if(!..())
+		return FALSE
+	for(var/mob/living/M in oview(C))
+		return TRUE
```
What's this for?

Doxxmedearly replied:
Paranoia has you hallucinate a mob in view doing some emote. If there's no mob in view, it should fail the can_affect() check so it doesn't get added to the list of candidates. If it was picked it would do nothing.

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+
+
+
+/datum/hallucination/skitter
```
All this whitespace needs to be removed. It's all over this file.

Doxxmedearly replied:
Was done for readability. Will adjust

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+
+
+/datum/hallucination/pain				//Pain. Picks a random type of pain, and severity is based on their level of hallucination.
+/datum/hallucination/pain/start()
```
There's also inconsistent spacing all across the file here from the declaration to the first proc.

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+					holder.emote("me",1,"flinches.")
+					holder.eye_blurry += 9
+
+	holder.adjustHalLoss(min(holder.hallucination / 3, 25))		//always cause fake pain
```
Consider making this pick a limb then directly adding pain to that limb. `/obj/item/organ/external/proc/add_pain`

NonQueueingMatt commented on code/modules/mob/living/carbon/hallucination.dm:
```diff
@@ -0,0 +1,1102 @@
+	var/attacker = pick(attacker_candidates)
+	attacker_candidates -= attacker
+	if(prob(50))
+		to_chat(holder, SPAN_DANGER("[attacker] has hit [holder]!"))
```
```suggestion
		to_chat(holder, SPAN_DANGER("[attacker] has punched [holder]!"))
```

Doxxmedearly commented:
!wip

Doxxmedearly commented:
!review

NonQueueingMatt commented on code/modules/hallucinations/types/mirage.dm:
```diff
@@ -0,0 +1,188 @@
+/datum/hallucination/mirage/bleeding/generate_mirage()
+	if(prob(min(holder.hallucination, 80)))
+		var/image/I = image('icons/effects/blood.dmi', pick("mgibbl1", "mgibbl2", "mgibbl3", "mgibbl4", "mgibbl5", "mfloor6", "mfloor7"), layer = TURF_LAYER)
+		I.color = holder.species.blood_color
```
```suggestion
		I.color = holder.species?.blood_color
```
Species might be absent if it is a carbon.

Doxxmedearly replied:
so THAT'S how those question marks are used.

NonQueueingMatt commented on code/modules/hallucinations/types/powers.dm:
```diff
@@ -0,0 +1,165 @@
+			H.end()
+		return
+
+	if(stat)
```
Perhaps this should be an use_check_and_message.

NonQueueingMatt commented on code/modules/hallucinations/types/powers.dm:
```diff
@@ -0,0 +1,165 @@
+	var/mob/target = input("Whose mind do you wish to probe?") as null|anything in creatures
+	if(isnull(target))
+		return
+	if(target.stat)
```
ditto

NonQueueingMatt replied:
wait nvm

NonQueueingMatt commented on code/modules/hallucinations/types/powers.dm:
```diff
@@ -0,0 +1,165 @@
+
+	var/mob/target = input("Whose mind do you wish to probe?") as null|anything in creatures
+	if(isnull(target))
+		return
```
```suggestion
	if(!target)
		return
```

NonQueueingMatt commented on code/modules/hallucinations/types/powers.dm:
```diff
@@ -0,0 +1,165 @@
+			H.end()
+		return
+
+	if(stat)
```
ditto

NonQueueingMatt commented on code/modules/hallucinations/types/powers.dm:
```diff
@@ -0,0 +1,165 @@
+	if(!creatures.len)
+		return
+	var/mob/target = input("Who do you wish to send a message to?") as null|anything in creatures
+	if(isnull(target))
```
```suggestion
	if(!target)
```

NonQueueingMatt commented on code/modules/hallucinations/types/powers.dm:
```diff
@@ -0,0 +1,165 @@
+	var/list/creatures = list()
+	for(var/mob/living/C in oview(usr))
+		creatures += C
+	creatures -= usr
```
oview already excludes the center

NonQueueingMatt commented on code/modules/hallucinations/types/powers.dm:
```diff
@@ -0,0 +1,165 @@
+		sleep(rand(50, 120))
+		to_chat(usr, SPAN_NOTICE("<b>You skim thoughts from the surface of \the [target]'s mind: \"<i>[pick(SShallucinations.hallucinated_phrases)]</i>\"</b>"))
+		for(var/mob/living/carbon/human/M in oviewers(src))
+			to_chat(M, "<B>[usr]</B> puts [usr.get_pronoun(1)] hands to [usr.get_pronoun(1)] head and mumbles incoherently as they stare, unblinking, at \the [target].")
```
I am pretty sure these can be made into a single visible_message? 
`/mob/visible_message(var/message, var/self_message, ...)`
There's also more occurrances of this around, so fix those too.

NonQueueingMatt commented on code/modules/hallucinations/types/powers.dm:
```diff
@@ -0,0 +1,165 @@
+	if(do_after(usr, 30))
+		to_chat(usr, SPAN_NOTICE("<b>You feel your message enter \the [target]'s mind!</b>"))
+		for(var/mob/living/carbon/human/M in oviewers(src))
+			to_chat(M,"<span class='game say'><span class='name'>[usr]</span> [usr.say_quote(message)], <span class='message'><span class='body'>\"[message]\"</span></span></span>")
```
Ditto on the visible_message

NonQueueingMatt commented on code/modules/hallucinations/hallucinations.dm:
```diff
@@ -0,0 +1,55 @@
+	var/allow_duplicates = TRUE		//This is set to false for hallucinations with long durations or ones we do not want repeated for a time
+	var/duration = 10				//how long before we call end()
+	var/min_power = 0 				//mobs only get this hallucination at this threshold
+	var/max_power = INFINITY		//mobs don't get this hallucination if it's above this threshold. Used to weed out more common ones if you're super fucked up
```
It might be worth making some basic defines for max_power and min_power. So something like this:
`HALLUCINATION_POWER_TINY 10` etc.

Arrow768 commented on html/changelogs/doxxmedearly - hallucination_overhaul.yml:
```diff
@@ -0,0 +1,24 @@
+  - rscadd: "Some hallucinations only appear if you're hallucinating heavily. There will be a difference between 'kinda hallucinating' and 'mind machine broke.'"
+  - rscadd: "Added several large lists of prompts that should keep hallucinations varied and interesting for a long time. These can also easily be expanded upon!"
+  - tweak: "Sounds will have less reverb while hallucinating, so you probably won't need to tear your headset off while resisting a cult conversion."
+  - backend: "Created a subsystem for hallucinations. Larger prompt lists were added as text files."
```
Please condense the changelog down to something that is shorter and more on point.

Doxxmedearly commented:
DNM until I can apply stylistic changes as requested. Thanks for the heads up!

Doxxmedearly commented:
Style updated, still need to do some final testing and we should be good to go.

Doxxmedearly commented:
fixed some typo issues with organs.
removed some weird holdover from old hallucination code that tried to emulate a fake shock but failed horribly.
made pain specify external organs.
removed the rage hallucination because frankly it was too dominant for a character and more annoying than anything.
removed some drowsy/eye blurry effects because they were more annoying than anything. 
allowed more time between effects to give people chances to respond to them. 
hopefully that's the last of the edits. I tested a few more times and things work; I'll just have to fine tune it after I see how people feel about it after it's been in-game for a bit.

skull132 commented:
Was test merged for about 3 days without issues. IN WE GO.
